### DISCUSSION TITLE
How to rebuild(rerender) page each day once at some specified time using getStaticProps?

### DISCUSSION THREAD

**Discussion by gadisauskas** [vercel/next.js#22803] [Help] [answered]
I'm using getServerSideProps to get data from back-end. Back-end data that I need for current page changes each day once at 4UTC. I want to implement getStaticProps which will rebuild all pages once each day, because data on server changes only once each day. I know about revalidate, but making revalidate with 24 hours also revalidates after each request. Also revalidating 24 hours is not exactly at 4UTC. So, how to rebuild page each day once at some specified time using getStaticProps function?

**Comment by ErnestBrandi** [comment]
Hi, I guess you could use a cron type service (like AWS CloudWatch events) in order to crawl (then revalidate) those pages at a certain time.

**Comment by lukeshumard** [comment, marked as answer]
Next.js doesn't currently offer a way to revalidate a page on-demand or programatically. There's a fair amount of discussion around that topic in #11552 if you're interested in reading more on the topic and other use cases.

## Solution 1: Revalidate more frequently
One solution might be to revalidate at a much more frequent interval than 24 hours, like once per minute. Even though the data only changes once per day, you can ensure that the data is "more up to date" using this method.

## Solution 2: Use a separate CDN
If revalidating a page on-demand is absolutely required, you can always set up your CDN outside of Next.js with a CDN that does allow revalidating on-demand such as AWS Cloudflare, Fastly, Cloudflare, etc. Your Next.js page would then use `getServerProps` and be rendered on-demand only if the CDN revalidates that page.

**Reply by gadisauskas** [reply]
I believe you wanted to say `getStaticProps` for solution 2. Because rebuilding app from CDN at 4UTC, the `getStaticProps` will be called and generates htmls for 1 day. Users will have fast pages downloads. Am I wrong? Because I can't imagine way, to make `getServerSideProps` be called from CDN and htmls be generated and placed in dist.

**Reply by lukeshumard** [reply]
I mean `getServerSideProps` for Solution 2. Let me try and explain a little better why I am suggesting this method.

When you use `getStaticProps`, Next.js will create a static output for that page so that it is not rendered on-demand. The big advantage with this approach is that you are not computing on-demand, which makes the response quicker but at the penalty of not necessarily having the freshest data.

When you use `getServerSideProps`, each request to that page is rendered on-demand. This means that `getServerSideProps` will always display the freshest data as it is pulling directly from whatever backend you are using, but at the penalty of being much slower than `getStaticProps`.

To combat the penalty of stale data while leveraging the performance gain of using `getStaticProps`, Next.js offers the `revalidate` option. This allows for your page to be regenerated at a certain interval (ie 10 seconds) but not a timestamp (ie 12:00 every day). By using a low interval for `revalidate`, this means your static page will get regenerated often to offer fresher data than every 24 hours. This is what I am proposing in **Solution 1**.

While the `revalidate` solution works for a lot of use cases, it's still not possible at this time to tell Next.js to regenerate a static page programatically. This is where using a separate CDN in front of Next.js would help. For example, you could use Cloudfront as your CDN with a Vercel-hosted Next.js app as your Origin. Your Next.js page uses `getServerSideProps` to render the freshest data at the time of request. Cloudfront caches the result of this request and serves that result. Once it reaches your desired time to regenerate, you can issue an Invalidation to Cloudfront to clear the cache for that page. This means the next request will hit your Origin and use `getServerSideProps` to get the freshest data. This is what I am proposing in **Solution 2**.

TL;DR - if it's okay to risk being a tiny bit stale, use **Solution 1** with `getStaticProps`. This involves much less setup and is handled directly by Next.js. I would highly recommend **Solution 1** as it is much easier to deal with. If you absolutely need to have programatic revalidation, use **Solution 2**. It requires more setup outside of Next.js, but allows you to fine-tune cache revalidation.
